src/Stack/Dot.hs

```diff
@@ -255,37 +255,57 @@ createDepLoader :: HasEnvConfig env
-                pkg <- loadCommonPackage (ppCommon pp)
-                pure (packageAllDeps pkg, payloadFromLocal pkg)
-              Just dp@DepPackage{dpLocation=PLImmutable loc} -> do
-                let common = dpCommon dp
-                gpd <- liftIO $ cpGPD common
-                let PackageIdentifier name version = PD.package $ PD.packageDescription gpd
-                    flags = cpFlags common
-                    ghcOptions = cpGhcOptions common
-                assert (pkgName == name) (loadPackageDeps pkgName version loc flags ghcOptions)
-              Nothing ->
-                pure (Set.empty, payloadFromInstalled (Map.lookup pkgName installed))
-      -- For wired-in-packages, use information from ghc-pkg (see #3084)
-      else case Map.lookup pkgName globalDumpMap of
-          Nothing -> error ("Invariant violated: Expected to find wired-in-package " ++ packageNameString pkgName ++ " in global DB")
-          Just dp -> pure (Set.fromList deps, payloadFromDump dp)
-            where
-              deps = map (\depId -> maybe (error ("Invariant violated: Expected to find " ++ ghcPkgIdString depId ++ " in global DB"))
-                                          Stack.Prelude.pkgName
-                                          (Map.lookup depId globalIdMap))
-                         (dpDepends dp)
+createDepLoader sourceMap installed globalDumpMap globalIdMap loadPackageDeps pkgName = do
+  case projectPackageDeps <|> dependencyDeps <|> globalDeps of
+    Just x -> x
+    Nothing ->
+      if isWiredInPackage then
+        let errText = "Invariant violated: Expected to find wired-in-package "
+        in error (errText ++ packageNameString pkgName ++ " in global DB")
+      else
+        pure noDeps
   where
+    isWiredInPackage = pkgName `Set.member` wiredInPackages
+    projectPackageDeps =
+      fmap loadDeps $ Map.lookup pkgName (smProject sourceMap)
+      where
+        loadDeps pp = do
+          pkg <- loadCommonPackage (ppCommon pp)
+          pure (packageAllDeps pkg, payloadFromLocal pkg)
+
+    dependencyDeps =
+      fmap loadDeps $ Map.lookup pkgName (smDeps sourceMap)
+      where
+        loadDeps dp =
+          case dp of
+            DepPackage{dpLocation=PLMutable dir} -> do
+              pp <- mkProjectPackage YesPrintWarnings dir False
+              pkg <- loadCommonPackage (ppCommon pp)
+              pure (packageAllDeps pkg, payloadFromLocal pkg)
+            DepPackage{dpLocation=PLImmutable loc} -> do
+              let common = dpCommon dp
+              gpd <- liftIO $ cpGPD common
+              let PackageIdentifier name version = PD.package $ PD.packageDescription gpd
+                  flags = cpFlags common
+                  ghcOptions = cpGhcOptions common
+              assert (pkgName == name) (loadPackageDeps pkgName version loc flags ghcOptions)
+
+    -- If package is a wired-in-package or a global package, use info from ghc-pkg (#4324, #3084)
```

Comment:
I'm not sure that "or" here is required - wired-in packages are supposed to always come as global and we don't deal with them at the level of this new `globalDeps` function anymore, so I think it makes sense to leave only global ones in this comment